File: fit_main.c

#include "nlopt.h"
#include <stdio.h>
#include <stdlib.h>


double opt_me(int n, const double *x, double *grad, void *func_data)
{
	const double a = x[0];
	const double b = x[1];
	const double v1 = (a*a + b - 11);
	const double v2 = (a + b*b - 7);

	return v1*v1 + v2*v2;
}


int main() {

	int maxeval = 1000;
	double minrms = 0.01;
	double tol = 0.0001;

	double lower[] = {-5., -5.};
	double upper[] = {5., 5.};
	double param_values[] = {0., 0.};
	double minf;

	//nlopt_opt opt = nlopt_create(NLOPT_G_MLSL_LDS, 2);
	nlopt_opt opt = nlopt_create(29, 2);
	nlopt_set_local_optimizer(opt, nlopt_create(NLOPT_LN_BOBYQA, 2));

	nlopt_set_lower_bounds(opt, lower);
	nlopt_set_upper_bounds(opt, upper);

	nlopt_set_min_objective(opt, opt_me, NULL);
	nlopt_set_maxeval(opt, maxeval);
	nlopt_set_stopval(opt, minrms);
	nlopt_set_ftol_abs(opt, tol);

	if (nlopt_optimize(opt, param_values, &minf) < 0)
		printf("NLopt failed!\n");
	else
		printf("Found minimum at f(%g,%g) = %0.10g\n",
		       param_values[0], param_values[1], minf);

	nlopt_destroy(opt);

	return 0;
}

File: fit_main.py
# Copyright (C) 2018 AGUILAR-MENA Jimmy and PANADES-BARRUETA Ramon L.
# This file is part of capi_nlopt <https://github.com/Panadestein/capi_nlopt.git> 
#
# capi_nlopt is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# capi_nlopt is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with capi_nlopt.  If not, see <http://www.gnu.org/licenses/>.
#!/usr/bin/env python3

import numpy as np
import wnlopt

# We'll optimize Himmelblau's function
def opt_me(x, grad):
    a, b = x[0], x[1]
    return (a**2 + b - 11)**2 + (a + b**2 - 7)**2

# Set variables
maxeval = 1000
minrms = 0.01
tol = 0.0001
param_values = np.array([0, 0], dtype=np.float64)

print(wnlopt.NLOPT_G_MLSL_LDS, wnlopt.NLOPT_LN_BOBYQA)

# C API wrapper
opt = wnlopt.PyNlopt(wnlopt.NLOPT_G_MLSL_LDS, 2)
opt.set_local_optimizer(wnlopt.PyNlopt(wnlopt.NLOPT_LN_BOBYQA, 2))

opt.set_lower_bounds(np.array([-5, -5], dtype=np.float64))
opt.set_upper_bounds(np.array([5, 5], dtype=np.float64))

opt.set_callback(opt_me)
opt.set_maxeval(maxeval)
opt.set_stopval(minrms)
opt.set_ftol_abs(tol)
minf = opt.optimize(param_values, 0.0)
print("minimum: f(%lf, %lf) = %lf" %
      (param_values[0], param_values[1], minf))
